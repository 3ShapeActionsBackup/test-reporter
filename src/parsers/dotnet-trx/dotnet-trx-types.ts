export interface TrxReport {
  TestRun: TestRun
}

export interface TestRun {
  Times: Times[]
  Results: Results[]
  TestDefinitions: TestDefinitions[]
}

export interface Times {
  $: {
    creation: Date
    queuing: Date
    start: Date
    finish: Date
  }
}

export interface TestDefinitions {
  UnitTest: UnitTest[]
}

export interface UnitTest {
  $: {
    id: string
  }
  TestMethod: TestMethod[]
}

export interface TestMethod {
  $: {
    className: string
    name: string
  }
}

export interface Results {
  UnitTestResult: UnitTestResult[]
}

export interface UnitTestResult {
  $: {
    testId: string
    testName: string
    duration: number
    outcome: Outcome
  }
  Output: Output[]
}

export interface Output {
  ErrorInfo: ErrorInfo[]
}
export interface ErrorInfo {
  Message: string[]
  StackTrace: string[]
}

export type Outcome = 'Passed' | 'NotExecuted' | 'Failed'
