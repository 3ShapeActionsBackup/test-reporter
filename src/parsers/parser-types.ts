import {TestRunResult} from '../report/test-results'

export type Annotation = {
  path: string
  start_line: number
  end_line: number
  start_column?: number
  end_column?: number
  annotation_level: 'notice' | 'warning' | 'failure'
  message: string
  title?: string
  raw_details?: string
}

export type ParseTestResult = (files: FileContent[], options: ParseOptions) => Promise<TestResult>

export type FileContent = {path: string; content: string}

export interface ParseOptions {
  annotations: boolean
  workDir: string
  trackedFiles: string[]
}

export interface TestResult {
  testRuns: TestRunResult[]
  annotations: Annotation[]
}
