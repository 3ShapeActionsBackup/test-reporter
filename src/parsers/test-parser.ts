import {Endpoints} from '@octokit/types'

type OutputParameters = Endpoints['POST /repos/:owner/:repo/check-runs']['parameters']['output']

export type ParseTestResult = (content: string) => Promise<TestResult>

export interface TestResult {
  success: boolean
  output: OutputParameters
}
