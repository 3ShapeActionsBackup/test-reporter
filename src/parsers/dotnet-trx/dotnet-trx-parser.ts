import {ErrorInfo, Outcome, TestMethod, TrxReport} from './dotnet-trx-types'

import {Annotation, ParseOptions, TestResult} from '../parser-types'
import {parseStringPromise} from 'xml2js'

import {parseAttribute} from '../../utils/xml-utils'
import {Icon} from '../../utils/markdown-utils'

import {
  TestExecutionResult,
  TestRunResult,
  TestSuiteResult,
  TestGroupResult,
  TestCaseResult
} from '../../report/test-results'
import getReport from '../../report/get-report'

class TestClass {
  constructor(readonly name: string) {}
  readonly tests: Test[] = []
}

class Test {
  constructor(
    readonly name: string,
    readonly outcome: Outcome,
    readonly duration: number,
    readonly error?: ErrorInfo
  ) {}

  get result(): TestExecutionResult {
    switch (this.outcome) {
      case 'Passed':
        return 'success'
      case 'NotExecuted':
        return 'skipped'
      case 'Failed':
        return 'failed'
    }
  }
}

export async function parseDotnetTrx(content: string, options: ParseOptions): Promise<TestResult> {
  const trx = (await parseStringPromise(content, {
    attrValueProcessors: [parseAttribute]
  })) as TrxReport

  const testClasses = getTestClasses(trx)
  const testRun = getTestRunResult(trx, testClasses)
  const success = testRun.result === 'success'
  const icon = success ? Icon.success : Icon.fail

  return {
    success,
    output: {
      title: `${options.name.trim()} ${icon}`,
      summary: getReport(testRun),
      annotations: options.annotations
        ? getAnnotations(/*testClasses, options.workDir, options.trackedFiles*/)
        : undefined
    }
  }
}

function getTestRunResult(trx: TrxReport, testClasses: TestClass[]): TestRunResult {
  const times = trx.TestRun.Times[0].$
  const totalTime = times.finish.getTime() - times.start.getTime()

  const suites = testClasses.map(tc => {
    const tests = tc.tests.map(t => new TestCaseResult(t.name, t.result, t.duration))
    const group = new TestGroupResult(null, tests)
    return new TestSuiteResult(tc.name, [group])
  })

  return new TestRunResult(suites, totalTime)
}

function getTestClasses(trx: TrxReport): TestClass[] {
  const unitTests: {[id: string]: TestMethod} = {}
  for (const td of trx.TestRun.TestDefinitions) {
    for (const ut of td.UnitTest) {
      unitTests[ut.$.id] = ut.TestMethod[0]
    }
  }

  const unitTestsResults = trx.TestRun.Results.flatMap(r => r.UnitTestResult).flatMap(unitTestResult => ({
    unitTestResult,
    testMethod: unitTests[unitTestResult.$.testId]
  }))

  const testClasses: {[name: string]: TestClass} = {}
  for (const r of unitTestsResults) {
    let tc = testClasses[r.testMethod.$.className]
    if (tc === undefined) {
      tc = new TestClass(r.testMethod.$.className)
      testClasses[tc.name] = tc
    }
    const output = r.unitTestResult.Output
    const error = output?.length > 0 && output[0].ErrorInfo?.length > 0 ? output[0].ErrorInfo[0] : undefined
    const test = new Test(r.testMethod.$.name, r.unitTestResult.$.outcome, r.unitTestResult.$.duration, error)
    tc.tests.push(test)
  }

  const result = Object.values(testClasses)
  result.sort((a, b) => a.name.localeCompare(b.name))
  for (const tc of result) {
    tc.tests.sort((a, b) => a.name.localeCompare(b.name))
  }

  return result
}

function getAnnotations(/*testClasses: TestClass[], workDir: string, trackedFiles: string[]*/): Annotation[] {
  return []
}
