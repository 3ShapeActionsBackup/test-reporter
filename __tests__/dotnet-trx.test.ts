import * as fs from 'fs'
import * as path from 'path'

import {parseDotnetTrx} from '../src/parsers/dotnet-trx/dotnet-trx-parser'
import {ParseOptions} from '../src/parsers/parser-types'

const xmlFixture = fs.readFileSync(path.join(__dirname, 'fixtures', 'dotnet-trx.trx'), {encoding: 'utf8'})
const outputPath = __dirname + '/__outputs__/dotnet-trx.md'

describe('dotnet-trx tests', () => {
  it('matches report snapshot', async () => {
    const opts: ParseOptions = {
      name: 'Dotnet TRX tests',
      annotations: true,
      trackedFiles: ['DotnetTests.Unit/Calculator.cs', 'DotnetTests.XUnitTests/CalculatorTests.cs'],
      workDir: 'C:/Users/Michal/Workspace/dorny/test-check/reports/dotnet/'
    }

    const result = await parseDotnetTrx(xmlFixture, opts)
    fs.mkdirSync(path.dirname(outputPath), {recursive: true})
    fs.writeFileSync(outputPath, result?.output?.summary ?? '')

    expect(result.success).toBeFalsy()
    expect(result?.output).toMatchSnapshot()
  })
})
