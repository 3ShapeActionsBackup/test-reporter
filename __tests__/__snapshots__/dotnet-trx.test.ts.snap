// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dotnet-trx tests matches report snapshot 1`] = `
Object {
  "annotations": Array [],
  "summary": "**7** tests were completed in **1.061s** with **3** passed, **1** skipped and **3** failed.
| Result | Suite | Tests | Time | Passed ✔️ | Failed ❌ | Skipped ✖️ |
| :---: | :--- | ---: | ---: | ---: | ---: | ---: |
| ❌ | [DotnetTests.XUnitTests.CalculatorTests](#ts-0-DotnetTests-XUnitTests-CalculatorTests) | 7 | 109.5761ms | 3 | 3 | 1 |
# Test Suites

## <a id=\\"user-content-ts-0-DotnetTests-XUnitTests-CalculatorTests\\" href=\\"#ts-0-DotnetTests-XUnitTests-CalculatorTests\\">DotnetTests.XUnitTests.CalculatorTests</a> ❌

| Result | Test | Time |
| :---: | :--- | ---: |
| ❌ | Exception_In_TargetTest | 0.4975ms |
| ❌ | Exception_In_Test | 2.2728ms |
| ❌ | Failing_Test | 3.2953ms |
| ✔️ | Passing_Test | 0.1254ms |
| ✔️ | Passing_Test_With_Name | 0.103ms |
| ✖️ | Skipped_Test | 1ms |
| ✔️ | Timeout_Test | 102.2821ms |
",
  "title": "Dotnet TRX tests ❌",
}
`;
